### PR TITLE
Automatically detect GPU architecture/code flags during CMake configure step

### LINKED ISSUE #18: Simplify specifying CUDA architectures in build system
Ginkgo's build system doesn't specify the GPU architectures on nvcc's command line. For this reason it just compiles the kernels to an intermediate code, which requires a jitter pass when the code is first run (maybe even every time the system is rebooted?), and even fails to build with CUDA 8 or lower, as it tries to build for c.c. 2.x, which is not supported by Ginkgo. To fix this the user has to manually modify the `CMAKE_CUDA_FLAGS` variable to set the architectures to the correct versions (this is awfully verbose if setting multiple architectures).

To simplify the process we could provide the user with a cmake variable which will let them set the architectures in a simpler way, the build system could then check if all the specified architectures are supported by Ginkgo (and throw a nice error if this is not true), and even detect the hardware and populate the variable automatically if the user doesn't set it by themselves.

### PR DESCRIPTION
**Pull request by yhmtsai**
Add CUDA_ARCH_NAME in Cmakefile for selecting how to decide arch

Auto: Auto detect GPU arch
Kepler, Maxwell, Pascal, Volta: Specify GPU
Manual: User sets sm_XX arch list (CUDA_ARCH_BIN)
All: all arch defined by nvcc version

This closes #18

### CODE REVIEW

**Comment by gflegar** [comment]
This was taken from [here](https://github.com/BVLC/caffe/blob/master/cmake/Cuda.cmake), correct?

Please always notify us if you are adding code from another project, this might not always be a problem, but we do have to check the licence of that project to see whether we're allowed to use their code. For example, it seems that caffe is distributed under the 2-clause BSD, which does require us to include their licence with the code.

I would suggest we put this part about architecture detection in a separate cmake file in the `cmake/` folder, and add caffe's licence on top of the file. Maybe add a comment stating that this file is taken from there, add a link to the project, and copy their licence. @hartwiganzt, @tcojean what do you think about this solution?

**Comment by tcojean** [comment]
Good catch Goran. We also maybe need to ask them for permission although it's not written in their License, just for politeness. Otherwise, I guess adding their licence on top and mention where it comes from should be enough.

In addition, only the last commit is relevant to this MR right? What is the merge order in respect to the other MRs because these 69 commits look scary.

**Comment by tcojean** [comment]
I would also suggest adding back the comments they had defined for the functions, because it's always a nice thing to have.

**Comment by gflegar** [comment]
My guess is that we should just copy the original file they have, add the licence and modify it to suit our needs. We can ask them for permission, but I'm not sure when (and if) we're going to get an answer - maybe we should go ahead and add it, ask them about it, and then deal with any requests from them once (and if) they reply.

Btw, it's possible @yhmtsai has taken it from a different project, as it seems that versions of this file are floating around github, for example [here](https://github.com/torch/cunn/pull/151/files).

**Comment by gflegar** [comment]
@tcojean about the 69 commits, I squashed the commits when merging some previous changes, as there was a lot of commits which first added something, and them removed the same thing, so the history didn't look nice. Since I don't want to tell external contributors how they should write and commit their code, this was the easiest solution. I guess they didn't update their fork with the newest changes in ginkgo, so that's why we're getting all these commits here. Most of the changes in them are already merged in the repo through several _big_ commits.

**Comment by yhmtsai** [comment]
@gflegar , sorry about that.
The reference code is the first one [(this)](https://github.com/BVLC/caffe/blob/master/cmake/Cuda.cmake)

For now, the following things are what I changed.
Add condition for the nvcc-8.0 and Volta and Kelper.
According the nvcc version to set up all arch.

The original one use `CUDA_NVCC_FLAGS`, but I did not find how they add the command into compile.
Will cmake use `CUDA_NVCC_FLAGS` automatically?
I am not sure whether it is a conflict with user's setting.
Thus, I use a cache to store this information and add command in gpu cmakefile.

**Comment by yhmtsai** [comment]
In the last 2-4 commit, I merge the last changed content from ginkgo.
But I do not know why it still calculates the commit from the first PR

**Comment by tcojean** [comment]
No it does not if you do a simple rebase, most likely it would also rebase all the extra commits which add nothing except changing the history. If you really want to clean it up yourself, I think the correct solution is an interactive rebase where you delete all commits except for the relevant one.

**Comment by tcojean** [comment]
@gflegar thanks for the explanation, indeed there is no change added so that means all the commits are already in the repo as you stated, but the history differs so they are kept.

**Comment by yhmtsai** [comment]
@gflegar 
Could I do these things first?
1. Reset the file to original ones
2. Copy Caffe's file, add their copyright in the file, mention where it is taken.
3. modify it.
Is there anything else I should do first?
Thanks.

**Comment by gflegar** [comment]
@yhmtsai sorry for not answering sooner - I was doing a couple of trips the last few weeks, so wasn't very active here.

Yes, you got it right, you can proceed with what you have proposed.

**Comment by yhmtsai** [comment]
A modification is done.

File: cmake/detect_gpu.cmake [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+    set(__archs_name_default "Auto")
+    if(NOT CMAKE_CROSSCOMPILING)
+        list(APPEND __archs_names "Auto")
+        set(__archs_name_default "Auto")
+    endif()
+
+    # set CUDA_ARCH_NAME strings (so it will be seen as dropbox in CMake-Gui)
+    set(CUDA_ARCH_NAME ${__archs_name_default} CACHE STRING
+        "Select target NVIDIA GPU achitecture.")
+    set_property( CACHE CUDA_ARCH_NAME PROPERTY STRINGS "" ${__archs_names} )
```
Comment (gflegar) [inline cmake/detect_gpu.cmake]: So what if I want to compile for multiple architectures? For example, both `Kepler` and `Pascal`. I cannot do that here, as I can only set 1 architecture name, and not a list of them.

Also, I might want to do: `Maxwell;Auto`, to compile for `Maxwell`, and anything else that was automatically detected. Or specifically, for all Maxwells, and 6.0 (maybe by setting `Maxwell;60`).
I also don't see the need for `Manual`, why wouldn't the user just set one variable to `60` instead of having to set two variables (one to `Manual`, and the other to `60`)?

This can be elegantly solved by looping through the input list of values, and just copying architectures to the output list, unless we find a keyword like `Kepler`, which is expanded to a list of `Kepler` architectures, or `Auto` which is expanded to a list of automatically detected architectures.

File: cmake/detect_gpu.cmake [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+    elseif(${CUDA_ARCH_NAME} STREQUAL "Maxwell")
+        set(__cuda_arch_bin "50 52")
+    elseif(${CUDA_ARCH_NAME} STREQUAL "Pascal")
+        set(__cuda_arch_bin "60 61")
+    elseif(${CUDA_ARCH_NAME} STREQUAL "Volta")
+        set(__cuda_arch_bin "70")
+    elseif(${CUDA_ARCH_NAME} STREQUAL "All")
+        set(__cuda_arch_bin ${ginkgo_known_gpu_archs})
+    elseif(${CUDA_ARCH_NAME} STREQUAL "Auto")
+        ginkgo_detect_installed_gpus(__cuda_arch_bin)
```
Comment (gflegar) [inline cmake/detect_gpu.cmake]: A more elegant solution, without the need for if's is to build an associative array:

```cmake
set(_cuda_arch_bin_Kepler ...)
set(_cuda_arch_bin_Maxwell ...)
set(_cuda_arch_bin_Pascal ...)
...
```
And then just do: `set(__cuda_arch_bin ${__cuda_arch_bin_${CUDA_ARCH_NAME}})`.

**Comment by yhmtsai** [comment]
The modification is done.
All options can be handled in CUDA_ARCH_OPTION
1. arch name (such as Fermi, Kepler...)
2. specified arch code can be BIN, BIN(PTX), or (PTX)
3. Auto: search all gpu arch in one station. (default)
4. All: all arch supported by CUDA
5. MaxPTX: it will embedding the max version PTX (default)

File: cmake/detect_gpu.cmake [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+
+
+################################################################################
+# Function for selecting GPU arch flags for nvcc based on CUDA_ARCH_OPTION
+# Usage:
+#     ginkgo_select_nvcc_arch_flags(out_variable)
+function(ginkgo_select_nvcc_arch_flags out_variable)
+    if(CMAKE_CUDA_COMPILER_VERSION MATCHES "([0-9]+).([0-9]+).(.*)")
+        set(cuda_version ${CMAKE_MATCH_1}${CMAKE_MATCH_2})
+    endif()
```
Comment (gflegar) [inline cmake/detect_gpu.cmake]: What if it doesn't match... we should return a `FATAL_ERROR` message that we cannot detect the compiler.

File: cmake/detect_gpu.cmake [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+        ${ginkgo_known_gpu_archs_name_${cuda_version}})
+    set(cuda_arch_bin_Volta ${cuda_arch_bin_Volta_${cuda_version}})
+    set(cuda_all_arch_list "")
+    foreach(__ver ${ginkgo_known_cuda_version})
+        list(APPEND cuda_all_arch_list ${ginkgo_known_gpu_archs_cuda_${__ver}})
+    endforeach()
+    list(REMOVE_DUPLICATES cuda_all_arch_list)
+
+    # set CUDA_ARCH_OPTION strings
+    set(CUDA_ARCH_OPTION "Auto" CACHE STRING
```
Comment (gflegar) [inline cmake/detect_gpu.cmake]: A minor nit: could we move this variable definition to the rest of the options in the top-level `CMakeListst.txt` file. This way all the options are defined in the same place, so users can quickly see which options we support.

File: cmake/detect_gpu.cmake [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+            check_available(${CMAKE_MATCH_1})
+            list(APPEND __cuda_arch_bin ${CMAKE_MATCH_1})
+        elseif(__option MATCHES "\\(([0-9]+)\\)")
+            check_available(${CMAKE_MATCH_1})
+            list(APPEND __cuda_arch_ptx ${CMAKE_MATCH_1})
+        elseif(__option STREQUAL "MaxPTX")
+            set(__bool_max_ptx "1")
+        elseif(__option STREQUAL "Auto")
+            ginkgo_detect_installed_gpus(__detect_bin)
+            check_available(${__detect_bin})
```
Comment (gflegar) [inline cmake/detect_gpu.cmake]: Maybe I'm missing something here, but from what I understand, `ginkgo_detect_installed_gpus()` will return a list of GPU architectures if more than 1 GPU is detected, while `check_available` expects only a single architecture as input.

Shouldn't check available be called in a loop over all entries in `__detect_bin`?

File: cmake/detect_gpu.cmake [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+#     ginkgo_select_nvcc_arch_flags(out_variable)
+function(ginkgo_select_nvcc_arch_flags out_variable)
+    if(CMAKE_CUDA_COMPILER_VERSION MATCHES "([0-9]+).([0-9]+).(.*)")
+        set(cuda_version ${CMAKE_MATCH_1}${CMAKE_MATCH_2})
+    endif()
+    set(ginkgo_known_gpu_archs ${ginkgo_known_gpu_archs_cuda_${cuda_version}})
+    set(ginkgo_known_gpu_archs_name
+        ${ginkgo_known_gpu_archs_name_${cuda_version}})
+    set(cuda_arch_bin_Volta ${cuda_arch_bin_Volta_${cuda_version}})
+    set(cuda_all_arch_list "")
```
Comment (gflegar) [inline cmake/detect_gpu.cmake]: nit: is this variable needed somewhere?

**Comment by gflegar** [comment]
> The original one use CUDA_NVCC_FLAGS, but I did not find how they add the command into compile.
> Will cmake use CUDA_NVCC_FLAGS automatically?
> I am not sure whether it is a conflict with user's setting.
> Thus, I use a cache to store this information and add command in gpu cmakefile.

@yhmtsai I just realized you asked this question - those few weeks when you asked it were a bit crazy for me, so I didn't read all the comments in detail. Sorry about that.

I'm no longer sure how the original code looked like, but the way we do it now is the correct way. `CUDA_NVCC_FLAGS` was a variable that the old [`FindCUDA`](https://cmake.org/cmake/help/v3.11/module/FindCUDA.html) CMake module used. This has now be deprecated since CUDA became a first-class language in CMake. When using the built-in CUDA support, CMake uses the `CMAKE_CUDA_FLAGS` variable for project-wide CUDA flags.

Again, this comment is just to answer your original question, there are no changes needed in the code because of it.

**Comment by hartwiganzt** [comment]
@gflegar : Works fine on Mac.

**Comment by gflegar** [comment]
@hartwiganzt  could you attache your CMakeCache.txt file? Thanks!

**Comment by gflegar** [comment]
It must have run it, that's the only way it could have gotten those flags. However, @yhmtsai it might be useful if the script also printed a message as soon as it detected the architectures, it's easier to debug if something goes wrong. Something like this output:

```
-- Detected GPU architectures: sm_60;sm_60
-- Added CUDA NVCC flags: sm_60 compute_60
```

The first line immediately tells us that automatic detection worked and found 2 GPUs, the second line tells us that it successfully combined all the flags and removed the duplicates.

File: cmake/detect_gpu.cmake (hunk 3 above)
Comment (yhmtsai) [reply cmake/detect_gpu.cmake]: got it.

File: cmake/detect_gpu.cmake (hunk 4 above)
Comment (yhmtsai) [reply cmake/detect_gpu.cmake]: move CUDA_ARCH_OPTION to top-level now

File: cmake/detect_gpu.cmake (hunk 5 above)
Comment (yhmtsai) [reply cmake/detect_gpu.cmake]: You are right. I added the foreach() to handle it now.

File: cmake/detect_gpu.cmake (hunk 6 above)
Comment (yhmtsai) [reply cmake/detect_gpu.cmake]: I use cuda_all_arch_list to verify whether arch is valid.

**Comment by yhmtsai** [comment]
I test it on our workstation (4 Titan V and 4 P100).
Some output of cmake.
```
-- Detected GPU architectures: 70;70;70;70;60;60;60;60
-- Added CUDA NVCC flags for: sm_70 sm_60 sm_72 sm_35 compute_72
```

However, it stuck at BlockJacobi.GpuPreconditionerEquivalentToRefWithMPW.
I also test BlockJacobi from the latest ginkgo with CMAKE_CUDA_FLAGS -arch=sm_35.
it also stuck at the corresponding testing.
(The cuda version is 9.1)

**Comment by gflegar** [comment]
@yhmtsai when you say it's stuck, you mean that the test freezes and doesn't continue?
On which GPU, the Titan V? I didn't have any problems with this on the P100 or my GTX965M. Didn't try it on a volta though...

File: cmake/detect_gpu.cmake (hunk 6 above)
Comment (gflegar) [reply cmake/detect_gpu.cmake]: Ah, I see now, could you the pass it as an argument to `check_variable`?
I know that `cmake` doesn't require you to do it, since its functions are macro-like constructs (somewhat similar to `#define function(...) { /*function body*/ }` in C and C++). But this is very unusual behavior and makes the code hard to follow. That's why I would prefer if we didn't rely on it.

**Comment by yhmtsai** [comment]
I fixed some typo in the previous comments.
Yes, the test freezes for at least 5 minutes on Titan V.

**Comment by gflegar** [comment]
About the test, we'll have to check what's going on at some point, but this is not a problem of this PR. Feel free to create an issue with your observations.

About this PR, I think it's pretty good now, I cloned the code and tested a few configurations, everything works like a charm.
